File: python/bob/bob.py
def response(phrase):
    if not phrase or phrase.isspace():
        return 'Fine. Be that way!'

    phrase_is_question = phrase.rstrip().endswith('?')

    if phrase.isupper():
        if phrase_is_question:
            return 'Calm down, I know what I\'m doing!'
        else:
            return 'Whoa, chill out!'

    if phrase_is_question:
        return 'Sure.'

    return 'Whatever.'

File: haskell/bob/src/Bob.hs
module Bob
  ( responseFor
  ) where

import Data.Char
import Data.String.Utils

responseFor :: String -> String
responseFor phrase
  | all isSpace phrase = "Fine. Be that way!"
  | isYell && isAsk = "Calm down, I know what I'm doing!"
  | isYell = "Whoa, chill out!"
  | isAsk = "Sure."
  | otherwise = "Whatever."
  where
    isYell =
      let content = filter isAlpha phrase
       in not (null content) && all isUpper content
    isAsk = endswith "?" $ rstrip phrase
